Add compatibility verifier scripts (#4854)

saileshnankani opened the pull request:
## Description

Adds 3 scripts to take 2 commit hashes and does rolling upgrade on the components to verify compatibility between commits.

- comp-verifier.sh (main script that does rolling upgrades) 
   command syntax: ./comp-verifier.sh commitHash1 target_dir_1 commitHash2 target_dir_2
- build-release.sh (helper script that builds Pinot given a specific commit hash and target directory )
   command syntax: ./build-release.sh commit_hash target_dir
- control-service.sh (helper script that starts/stops given component for a given build of Pinot)
   command syntax: ./control-service.sh command component target_dir pinot_version

Starts all components in v1 version. Afterwards, it does a rolling upgrade - controller first, then the broker, and finally the server to v2. 

See #4854 for design documents

mcvsubbu commented on compatibility-verifier/build-release.sh:
```diff
@@ -0,0 +1,59 @@
+
+# A helper script that builds Pinot given a specific commit hash and target directory
+# This script is intended to be called stand-alone or from comp-verifier.sh 
+# command syntax: ./comp-verifier.sh commitHash1 target_dir_1 commitHash2 target_dir_2
```
Let us use this script to build one release, and invoke it as many times as needed

mcvsubbu commented on compatibility-verifier/build-release.sh:
```diff
@@ -0,0 +1,59 @@
+# verify correct usage of this script 
+if [[ $# -ne 2 ]]; then
+    printf "You used %s arguments \n" "$#"
+    printf "Usage: \n ./build-release.sh commit_hash target_dir \n"
```
Introduce a function called usage() and print the usage in it. We can call that function from multiple points as necessary

mcvsubbu replied:
```suggestion
    echo "Usage:$0 commit_hash target_dir "
```

mcvsubbu commented on compatibility-verifier/build-release.sh:
```diff
@@ -0,0 +1,59 @@
+# command syntax: ./comp-verifier.sh commitHash1 target_dir_1 commitHash2 target_dir_2
+
+# verify correct usage of this script 
+if [[ $# -ne 2 ]]; then
```
Please use '[' instead of '[['. Won't repeat for other instances further down in this file and other files

mcvsubbu commented on compatibility-verifier/build-release.sh:
```diff
@@ -0,0 +1,59 @@
+
+# verify correct usage of this script 
+if [[ $# -ne 2 ]]; then
+    printf "You used %s arguments \n" "$#"
```
Remove this line

mcvsubbu commented on compatibility-verifier/build-release.sh:
```diff
@@ -0,0 +1,59 @@
+commit_hash=$1
+target_dir=$2
+
+# check if directory already exists and checkout  
```
exit if the directory exists. We can add a -f argument later or force

mcvsubbu commented on compatibility-verifier/build-release.sh:
```diff
@@ -0,0 +1,59 @@
+# check if directory already exists and checkout  
+# using the given commit
+if [[ ! -d $target_dir ]]; then
+    mkdir -p "$target_dir"
```
You will need to check the exit status of mkdir in case there were permission issues

mcvsubbu commented on compatibility-verifier/comp-verifier.sh:
```diff
@@ -0,0 +1,145 @@
+# verify correct usage of this script 
+if [[ $# -ne 4 ]]; then
+    printf "You used %s arguments \n" "$#"
+    printf "Usage: \n ./comp-verifier.sh commitHash1 target_dir_1 commitHash2 target_dir_2 \n"
```
Take a working dir as argument and mkdir two subdirs in it, one for each commit hash.

mcvsubbu replied:
```suggestion
    echo "Usage: $0 earlierCommitHash laterCommitHash workingDir"
```

mcvsubbu commented on compatibility-verifier/comp-verifier.sh:
```diff
@@ -0,0 +1,145 @@
+pinot_version_1="0.5.0"
+pinot_version_2="0.5.0"
+
+read -rep $'\n' -p "What is the Pinot version for first commitHash? [default: 0.5.0] " -r
```
Why read pinot versions? We have the commit hash, pick up whatever version or hash that is.

saileshnankani replied:
I required it because I'm using pinot-admin.sh to start the components and the path to pinot-admin.sh file is dependent on the Pinot version for e.g., `pinot-distribution/target/apache-pinot-incubating-0.5.0-SNAPSHOT-bin/apache-pinot-incubating-0.5.0-SNAPSHOT-bin/`. I'm not sure if I can get the Pinot version from the hash.

saileshnankani replied:
Never mind. I noticed there is also `./pinot-tools/target/pinot-tools-pkg/bin/pinot-admin.sh` that I can use.

mcvsubbu commented on compatibility-verifier/comp-verifier.sh:
```diff
@@ -0,0 +1,145 @@
+#
+
+# A script that does rolling upgrade of Pinot components
+# from one version to the other given 2 commit hashes. It first builds  
```
Mention clearly that the script assumes the first hash is the earlier one and the second is the later one. Change your variable names to indicate so

mcvsubbu commented on compatibility-verifier/comp-verifier.sh:
```diff
@@ -0,0 +1,145 @@
+  ./build-release.sh "$commitHash2" "$target_dir_2"
+fi
+
+if [[ $(lsof -t -i:7001 -s TCP:LISTEN) || $(lsof -t -i:8001 -sTCP:LISTEN) || $(lsof -t -i:9001 -sTCP:LISTEN) || 
```
Do not kill existing processes. If the ports are not open, exit with an error saying so
We will add additional a config file for each component to fine-tune things. For now,
default configs and ports are good enough to get things going.

mcvsubbu commented on compatibility-verifier/comp-verifier.sh:
```diff
@@ -0,0 +1,145 @@
+  ./build-release.sh "$commitHash2" "$target_dir_2"
+fi
+
+if [[ $(lsof -t -i:7001 -s TCP:LISTEN) || $(lsof -t -i:8001 -sTCP:LISTEN) || $(lsof -t -i:9001 -sTCP:LISTEN) || 
```
Not sure if lsof returns all ports even if you don't have permissions

saileshnankani replied:
I'm currently not sure how to circumvent this issue. I could try adding sudo, but that would prompt users with password.

mcvsubbu commented on compatibility-verifier/comp-verifier.sh:
```diff
@@ -0,0 +1,145 @@
+  fi
+fi
+
+read -rep $'\n' -p "Do you want to build the cluster with first commit? [default: no] " -n 1 -r
```
remove this prompt. The user has entered the command, and is therefore ready to build the cluster.

mcvsubbu commented on compatibility-verifier/comp-verifier.sh:
```diff
@@ -0,0 +1,145 @@
+
+# Setup initial cluster (with commit in target_dir_1)
+
+printf "\n\nStarting Zookeeper ...\n"
```
You don't need these prompts.

mcvsubbu commented on compatibility-verifier/control-service.sh:
```diff
@@ -0,0 +1,104 @@
+
+# A helper script that starts/stops given component for a given build of Pinot
+# This script is intended to be called stand-alone or from comp-verifier.sh 
+# command syntax: ./control-service.sh command component target_dir pinot_version
```
You don't need pinot version

mcvsubbu commented on compatibility-verifier/control-service.sh:
```diff
@@ -0,0 +1,104 @@
+fi
+
+# Define ports for the test
+zkPort=2181
```
For now, just use the default ports on the service. Let us get a basic framework working. and then we can add the bells and whistles.

mcvsubbu commented on compatibility-verifier/control-service.sh:
```diff
@@ -0,0 +1,104 @@
+  # Start the desired component
+  # Upon start, save the pid of the process for a component into a file in /tmp/{component}.pid, which is then used to stop it
+  if [[ "$component" == "controller" ]]; then
+    sh -c 'echo $$ > /tmp/controller.pid; exec ./pinot-admin.sh StartController -zkAddress localhost:$0 -clusterName PinotCluster -controllerPort $1 > /tmp/pinot-controller.log' ${localhost} "${controllerPort}"
```
Just start with default ports for now.

saileshnankani commented:
> Overall: Thanks for getting this started.
> 
> 1. You don't need to use version
> 2. Remove all prompts. The test should be push-button. Top-level goal is to enter a single command with two commit hashes (and maybe a working dir) and it should run all the tests we have.
> 3. Please use camel case instead of underscores
> 4. Use shell functions when possible. Adding a function for usage() has often been useful for me, to be invoked from anywhere we see discrepancies that we think are user errors.
> 
> High-Level command:
> `compatCheck.sh olderCommit newerCommit [workingDir]`
> Sub-commands that you will find useful to develop, either as functions or separate commands:
> `checkoutAndBuild commitHash dirName`
> `startServices dirName`
> `stopServices dirName`
> 
> For now, functions should be enough, we can split the files as needed later.

Do you suggest that startServices/stopServices functions should be available along with control-service subcommand or is it supposed to replace it? In the case it replaces the control-service file entirely, its command syntax suggests that we lose the ability to do a rolling upgrade as we don't have control over a specific component anymore.

mcvsubbu commented:
I fixed my comment. Hope that makes it clearer. Here is the general outline (as of now, and we can always refine it as we go along).
`compatCheck.sh` does the following:
1. Checkout and build in old-version
2. checkout and build in new version.
3. start  zookeeper in any one version (old or new, does not matter)
4. Start all services  in old version (controller first, broker/server in any order).
5. upgradeAnd Test()
6. dowgradeAndTest()

`upgradeAndTest()` does the following:
1. Stop controller in old version
2. Start controller in new version.
2a. More code here later
3. Stop broker in old version
4. Start broker in new version.
4a. More code in later
5. Stop server in old version
6. Start server in new version
6a. More code here later.

`downgradeAndTest()` does the folllowing:
1. Stop server in new version
2. Start server in old version
2a. More code to add here later
3. Stop broker in new version
4. Start broker in old version
4a. More code here later
5. Stop controller in new version
6. Start controlller in new version
6a. More code here later.

saileshnankani commented:
I have implemented most of the changes you've requested. There's an additional change I made:

- Since providing a working directory is optional, I added code for building Pinot in temporary directory and handling the cleanup at exit.

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+}
+
+# cleanp the temporary directory when the bash script exits 
+trap cleanup EXIT
```
Please move all the functions to the top and main code below the function defintions

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+
+# This function builds Pinot given a specific commit hash and target directory
+function checkoutAndBuild() {
+  if [ $# -ne 2 ]; then 
```
Since these are internal function calls, you don't need to check for usage here

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+
+# Given a component and directory, start that version of the specific component 
+function startService() {
+  if [ $# -ne 2 ]; then 
```
dont need usage check

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+  fi
+  serviceName=$1
+  dirName=$2
+  if [[ ! " ${components[*]} " =~ $serviceName ]]; then
```
dont need this check

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+
+# Given a component, check if it known to be running and stop that specific component
+function stopService() {
+  if [ $# -ne 1 ]; then 
```
remove

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+    usage stopService
+  fi
+  serviceName=$1
+  if [[ ! " ${components[*]} " =~ $serviceName ]]; then
```
remove

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+  # Upon start, save the pid of the process for a component into a file in /tmp/{component}.pid, which is then used to stop it
+  pushd "$dirName"/pinot-tools/target/pinot-tools-pkg/bin  || exit 1
+  if [ "$serviceName" = "zookeeper" ]; then
+    sh -c 'echo $$ > /tmp/zookeeper.pid; exec ./pinot-admin.sh StartZookeeper' &
```
create the pid file inside $dirName for the release that we are working with.

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+      kill -9 "$servicePid"
+    fi
+  else
+    echo "Can't stop component ${serviceName} since can't component ${serviceName} is not known to have been started."
```
add the full path name to pid file in this message. "Pid file {} not found. Failed to stop component {}"

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+  echo "Cluster stopped."
+} 
+
+# upgrades the cluster by upgrading components one by one (Controller -> Broker -> Server) 
```
We need to run tests before/after each step, so there will be a lot of code between each stop and start. Don't put this as a function as yet.. It can be in the main code for now.

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+fi
+
+# Building targets
+echo "Building the first target ... "
```
```suggestion
echo "Building old version "
```

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+# Building targets
+echo "Building the first target ... "
+checkoutAndBuild "$olderCommit" "$oldTargetDir"
+echo "Building the second target ..."
```
```suggestion
echo "Building new version"
```

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+fi
+
+# Setup initial cluster with olderCommit and do rolling upgrade
+startServices "$oldTargetDir"
```
unroll the switchcomponents function here

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,210 @@
+  startService server "$dirName"
+}
+
+# Do rolling upgrade from olderCommit to newerCommit
```
remove lines 167 to 171 for now, we will add tests later

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,173 @@
+
+# cleanup the temporary directory when exiting the script
+function cleanup() {
+  if [ -n "$tmpDir" ] && [ -d "$tmpDir" ]; then
```
For now, let us just echo that this needs to be cleaned up.
In some cases, we would want to keep the dir to see the logs/results, so it may be useful to keep it.
Also, I would prefer that we set a variable called TMP_DIR up front to the results of mktemp. The use of `TMP_DIR` indicates that we don't modify it -- just a convention.

saileshnankani replied:
So do you mean just renaming tmpDir to TMP_DIR?

mcvsubbu replied:
just move the assignment to `tmpDir` to be an assignment to `TMP_DIR` and move it to the top of the file. It does not create the dir anyway, only assigns a dir name. We can create the dir if needed (if working dir is not supplied)

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,173 @@
+    servicePid=$(<"${dirName}/${serviceName}".pid)
+    rm "${dirName}/${serviceName}".pid
+    if [ -n "$servicePid" ]; then
+      kill -9 "$servicePid"
```
Unfortunate, we need a better way to stop a service gracefully. For now, that is ok

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,173 @@
+# get usage of the script 
+function usage() {
+  command=$1
+  if [[ "$command" =~ compCheck.sh$ ]] ; then
```
Let us remove this if

mcvsubbu commented on compatibility-verifier/compCheck.sh:
```diff
@@ -0,0 +1,173 @@
+newTargetDir="$workingDir"/newTargetDir
+
+if ! mkdir -p "$oldTargetDir" "$newTargetDir"; then
+  echo "Failed to create target directory"
```
include the name of the directory that failed to create
